Enable CMPXCHG16B, SSE3, SAHF/LAHF and 128-bit Atomics (in nightly) in Windows x64

CKingX opened the pull request:
As Rust plans to set Windows 10 as the minimum supported OS for target x86_64-pc-windows-msvc, I have added the cmpxchg16b and sse3 feature. Windows 10 requires CMPXCHG16B, LAHF/SAHF, and PrefetchW as stated in the requirements [here](https://download.microsoft.com/download/c/1/5/c150e1ca-4a55-4a7e-94c5-bfc8c2e785c5/Windows%2010%20Minimum%20Hardware%20Requirements.pdf). Furthermore, CPUs that meet these requirements also have SSE3 ([see](https://walbourn.github.io/directxmath-sse3-and-ssse3/))

CKingX committed:
As Windows 10 requires certain features like CMPXCHG16B and a few others and Rust plans to set Windows 10 as the minimum supported OS for target x86_64-pc-windows-msvc, I have added the cmpxchg16b and sse3 feature (as CPUs that meet the Windows 10 64-bit requirement also support SSE3. See https://walbourn.github.io/directxmath-sse3-and-ssse3/ )

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7838356093/job/21389656302) [(plain)](https://github.com/rust-lang/rust/commit/d51e703534c49b5b658a954fe6d387c33ba0c5e3/checks/21389656302/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=CKingX
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_ca1af936-ef69-4fdb-a728-2363830ebf10
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=f353f8c81685bbc57e68d9812ea43bc56048d351
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_ca1af936-ef69-4fdb-a728-2363830ebf10
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_ca1af936-ef69-4fdb-a728-2363830ebf10
GITHUB_TRIGGERING_ACTOR=CKingX
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/120820/merge
GITHUB_WORKFLOW_SHA=f353f8c81685bbc57e68d9812ea43bc56048d351
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
#12 writing image sha256:196d8f6f3433f557e2ece20a1d14b59026ceedc0c92078a2c78f2ca016e6b29d done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Fri Feb  9 01:26:42 UTC 2024
  network time: Fri, 09 Feb 2024 01:26:42 GMT
  network time: Fri, 09 Feb 2024 01:26:42 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [ui] tests/ui/abi/compatibility.rs#x86-64-win stdout ----
normalized stderr:
'+cmpxchg16b' is not a recognized feature for this target (ignoring feature)
'+cmpxchg16b' is not a recognized feature for this target (ignoring feature)


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/abi/compatibility.x86-64-win/compatibility.x86-64-win.stderr
To only update this specific test, also pass `--test-args abi/compatibility.rs`


error in revision `x86-64-win`: 1 errors occurred comparing output.
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/abi/compatibility.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "x86_64_win" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/abi/compatibility.x86-64-win" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/abi/compatibility.x86-64-win/auxiliary" "--target" "x86_64-pc-windows-msvc"
--- stderr -------------------------------
--- stderr -------------------------------
'+cmpxchg16b' is not a recognized feature for this target (ignoring feature)
'+cmpxchg16b' is not a recognized feature for this target (ignoring feature)



failures:
```

</details>

CKingX commented:
Strange. `rustc --print=target-features` lists cmpxchg16b

ChrisDenton commented:
Probably because llvm disables it by default. There's likely some argument llvm needs to enable enabling the feature. But I'm not an LLVM expert.

ChrisDenton commented on compiler/rustc_target/src/spec/targets/x86_64_pc_windows_msvc.rs:
```diff
@@ -3,6 +3,7 @@ use crate::spec::{base, SanitizerSet, Target};
 pub fn target() -> Target {
     let mut base = base::windows_msvc::opts();
     base.cpu = "x86-64".into();
+    base.features = "+cmpxchg16b,+sse3".into();
```
`+sahf` too?

CKingX replied:
That I am not sure. x86-64-v2 onwards don't use sahf though it is listed in the features. CMPXCHG16B and SSE3 are listed. Does that mean Rustc doesn't yet make use of these instructions?

CryZe replied:
Not sure what you are seeing but x86-64-v2 totally does specify SAHF:
https://github.com/llvm/llvm-project/blob/a9700904765590ca2fbf08c0cc36d0da1107d3a7/llvm/lib/Target/X86/X86.td#L808-L809

CKingX replied:
I used this
```
rustc --print=cfg -C target-cpu=x86-64-v2
debug_assertions
panic="unwind"
target_arch="x86_64"
target_endian="little"
target_env="msvc"
target_family="windows"
target_feature="cmpxchg16b"
target_feature="fxsr"
target_feature="popcnt"
target_feature="sse"
target_feature="sse2"
target_feature="sse3"
target_feature="sse4.1"
target_feature="sse4.2"
target_feature="ssse3"
target_has_atomic="16"
target_has_atomic="32"
target_has_atomic="64"
target_has_atomic="8"
target_has_atomic="ptr"
target_os="windows"
target_pointer_width="64"
target_vendor="pc"
windows
```

Did some more digging and turns out SAHF seems to be for kernel rather than userspace code. Hence it is not listed in the rustc command I ran above.

CKingX replied:
It also seems to be tied to virtualization and for kernel space.

CryZe replied:
Ah seems like LLVM does indeed explicitly disable it for 64-bit targets: https://github.com/llvm/llvm-project/blob/7ddc32052546abd41656d2e670f3902b1bf805a7/llvm/lib/Target/X86/X86Subtarget.h#L228C40-L228C53

ChrisDenton replied:
Ok, I've been researching this a bit more. On balance I do think we should add `+sahf`. While I don't think it's that important, it was seen fit to be included in [x86-64-v2](https://en.wikipedia.org/wiki/X86-64#Microarchitecture_levels) feature level and I think we should align ourselves with that (although obviously dropping the SSE level back down to 3).

CKingX replied:
done!

ChrisDenton commented:
Hm, using `+cx16` appears to work but for some reason `cmpxchg16b` doesn't.

CKingX committed:
Update x86_64_pc_windows_msvc.rs

Fixed a bug where adding CMPXCHG16B would fail due to different names in Rustc and LLVM

CKingX commented:
Thank you! Fixed it

CKingX commented:
Also as CMPXCHG16B allows 128-bit/16-byte atomics, I am creating a new commit to update max atomic width as well. This is also already done for x86_64_apple_darwin target where it states:
`    base.max_atomic_width = Some(128); // penryn+ supports cmpxchg16b`

CKingX committed:
Update x86_64_pc_windows_msvc.rs

As CMPXCHG16B is supported, I updated the max atomic width to 128-bits from 64-bits

CKingX commented:
x86-64 darwin (macOS) target does this without AVX support as the minimum requirement is Penryn (Core 2) which doesn't support AVX or AVX2.

mati865 commented:
Shouldn't this be set for all other x86 Windows targets as well?

CKingX commented:
As for 32-bit Windows 10, it only requires SSE2 and that is already the requirement in Rustc.

Now for 64-bit MinGW ABI, I could update it if that also requires Windows 10 and up. I looked at Rust [platforms](https://doc.rust-lang.org/rustc/platform-support.html) and I didn't see win7 version of MinGW ABI so I am unsure. However, if the -gnu targets also dropped support, I can do a new commit

ChrisDenton commented:
All `x86_64-pc-windows` targets with std support are raising their baseline to Windows 10. Also the `uwp` targets because they inherently don't support older Windows versions. The only exceptional one is the win7 target.

CKingX committed:
Update x86_64_uwp_windows_gnu.rs

Updated x86_64-uwp-windows-gnu to use CMPXCHG16B and SSE3

CKingX committed:
Possibly removed merge policy

CKingX commented:
Done. Let me know if I missed any

fmease commented:
I'm not super familiar with this.
r? compiler

petrochenkov commented:
r=me once @ChrisDenton is happy with the specific set of added features.
r? @ChrisDenton

CKingX committed:
Added sahf feature to windows targets

ChrisDenton commented:
Looks good to me, thanks!

@bors r=petrochenkov

bors commented:
:pushpin: Commit 376c7b98921b427699b99fc36496cb81ed40c946 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 376c7b98921b427699b99fc36496cb81ed40c946 -->
<!-- homu: {"type":"Approved","sha":"376c7b98921b427699b99fc36496cb81ed40c946","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 376c7b98921b427699b99fc36496cb81ed40c946 with merge d6a09a74320b24ee8fc62d1190d844e367fb2ba6...
<!-- homu: {"type":"BuildStarted","head_sha":"376c7b98921b427699b99fc36496cb81ed40c946","merge_sha":"d6a09a74320b24ee8fc62d1190d844e367fb2ba6"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/7955720699/job/21715170234) [(plain)](https://github.com/rust-lang-ci/rust/commit/d6a09a74320b24ee8fc62d1190d844e367fb2ba6/checks/21715170234/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests\codegen\sse42-implies-crc32.rs stdout ----

error: verification with 'FileCheck' failed
status: exit code: 1
command: PATH=";C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64;C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage0-bootstrap-tools\x86_64-pc-windows-msvc\release\deps;C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage0\bin;C:\Program Files\PowerShell\7;C:\a\rust\rust\ninja;C:\a\rust\rust\msys2\mingw64\bin;C:\hostedtoolcache\windows\Python\3.12.2\x64\Scripts;C:\hostedtoolcache\windows\Python\3.12.2\x64;C:\msys64\usr\bin;C:\a\rust\rust\sccache;C:\Program Files\MongoDB\Server\5.0\bin;C:\aliyun-cli;C:\vcpkg;C:\cf-cli;C:\Program Files (x86)\NSIS;C:\tools\zstd;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\2.15.1\x64;C:\cabal\bin;C:\ghcup\bin;C:\mingw64\bin;C:\Program Files\dotnet;C:\Program Files\MySQL\MySQL Server 5.7\bin;C:\Program Files\R\R-4.3.2\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\SeleniumWebDrivers\ChromeDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.20.14\x64\bin;C:\hostedtoolcache\windows\Python\3.7.9\x64\Scripts;C:\hostedtoolcache\windows\Python\3.7.9\x64;C:\hostedtoolcache\windows\Ruby\2.5.9\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.402-6\x64\bin;C:\Program Files\ImageMagick-7.1.1-Q16-HDRI;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot\bin;C:\ProgramData\Chocolatey\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\Microsoft SQL Server\110\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\120\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\140\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\160\DTS\Binn;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\TortoiseSVN\bin;C:\Program Files\CMake\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.8.7\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files (x86)\Google\Cloud SDK\google-cloud-sdk\bin;C:\Program Files (x86)\Microsoft BizTalk Server;C:\Program Files\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\llvm\\build\\bin\\FileCheck.exe" "--input-file" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\codegen\\sse42-implies-crc32\\sse42-implies-crc32.ll" "C:\\a\\rust\\rust\\tests\\codegen\\sse42-implies-crc32.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,MSVC" "--dump-input-context" "100"
--- stderr -------------------------------
C:\a\rust\rust\tests\codegen\sse42-implies-crc32.rs:15:11: error: CHECK: expected string not found in input
C:\a\rust\rust\tests\codegen\sse42-implies-crc32.rs:15:11: error: CHECK: expected string not found in input
// CHECK: attributes #0 {{.*"target-features"="\+sse4.2,\+crc32"}}
          ^
C:\a\rust\rust\build\x86_64-pc-windows-msvc\test\codegen\sse42-implies-crc32\sse42-implies-crc32.ll:1:1: note: scanning from here
; ModuleID = 'sse42_implies_crc32.2382b1a2b42691d8-cgu.0'
^
C:\a\rust\rust\build\x86_64-pc-windows-msvc\test\codegen\sse42-implies-crc32\sse42-implies-crc32.ll:16:99: note: possible intended match here
attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "target-cpu"="x86-64" "target-features"="+cx16,+sse3,+sahf,+sse4.2,+crc32" }


Input file: C:\a\rust\rust\build\x86_64-pc-windows-msvc\test\codegen\sse42-implies-crc32\sse42-implies-crc32.ll
Check file: C:\a\rust\rust\tests\codegen\sse42-implies-crc32.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'sse42_implies_crc32.2382b1a2b42691d8-cgu.0' 
check:15'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: source_filename = "sse42_implies_crc32.2382b1a2b42691d8-cgu.0" 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            3: target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            4: target triple = "x86_64-pc-windows-msvc" 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            5:  
check:15'0     ~
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7: define noundef i32 @crc32sse(i8 noundef %v) unnamed_addr #0 { 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            8: start: 
check:15'0     ~~~~~~~
            9:  %_0 = tail call noundef i32 @llvm.x86.sse42.crc32.32.8(i32 noundef -1, i8 noundef zeroext %v) #2 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  ret i32 %_0 
check:15'0     ~~~~~~~~~~~~~
           11: } 
check:15'0     ~~
           12:  
check:15'0     ~
           13: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(none) 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14: declare i32 @llvm.x86.sse42.crc32.32.8(i32, i8) unnamed_addr #1 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           15:  
check:15'0     ~
           16: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "target-cpu"="x86-64" "target-features"="+cx16,+sse3,+sahf,+sse4.2,+crc32" } 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:15'1                                                                                                       ?                                                                           possible intended match
           17: attributes #1 = { mustprogress nocallback nofree nosync nounwind willreturn memory(none) } 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           18: attributes #2 = { nounwind } 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           19:  
check:15'0     ~
           20: !llvm.module.flags = !{!0} 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21: !llvm.ident = !{!1} 
check:15'0     ~~~~~~~~~~~~~~~~~~~~
           22:  
check:15'0     ~
           23: !0 = !{i32 8, !"PIC Level", i32 2} 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24: !1 = !{!"rustc version 1.78.0-nightly (d6a09a743 2024-02-19)"} 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------




failures:
    [codegen] tests\codegen\sse42-implies-crc32.rs

test result: FAILED. 424 passed; 1 failed; 150 ignored; 0 measured; 0 filtered out; finished in 6.17s

Some tests failed in compiletest suite=codegen mode=codegen host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
Build completed unsuccessfully in 0:41:00
make: *** [Makefile:73: ci-msvc-ps1] Error 1
  network time: Mon, 19 Feb 2024 07:44:51 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

CKingX commented:
Ok so the test is checking if SSE4.2 enables CRC32 on x86-64 (base). As we passed in extra features, it fails.

I am going to look into how other targets handle this when they add feature flags

CKingX commented:
I did some digging and it is not all good news. x86_64h-apple-darwin handles this by using core-avx2 and removing some features using the "-feature" syntax. Now, it evaluates the feature left to right so "-feature,+feature" should still lead to it correcting. The disadvantage of this approach is that it doesn't work if you change the target cpu (like choosing native) where those features don't get added back.

That leaves updating the test, but I am unsure on if the test should be changed.

CKingX commented:
The other idea is setting the target cpu as nocona but that might mess up optimizations as a whole because nocona is based off of pentium 4 (far off from modern CPUs as compared to core to haswell)

erikdesjardins commented:
From the test failure it looks like that line changes to

```ll
attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "target-cpu"="x86-64" "target-features"="+cx16,+sse3,+sahf,+sse4.2,+crc32" }
```

So it's fine to just change the [test expectation](https://github.com/rust-lang/rust/blob/0b9f6ad99482c9fc35baa5f40cebf21be4fd98b3/tests/codegen/sse42-implies-crc32.rs#L15C1-L15C67) to allow additional target features before sse4.2/crc32, like
```
// CHECK: attributes #0 {{.*"target-features"=".*\+sse4.2,\+crc32"}}
```
since that upholds the spirit of the test (that adding sse4.2 also adds crc32)

CKingX committed:
Updated test to account for added previous features (thanks erikdesjardins!)

CKingX commented:
Thanks erikdesjardins! Updating the test was a lot simpler than I expected.

CKingX commented:
Just as a heads up the updated test just needs to be reviewed

ChrisDenton commented:
lgtm, thanks!

@bors r=petrochenkov,ChrisDenton

bors commented:
:pushpin: Commit 2d25c3b3697728ea42eb26f79e6f80907470ca22 has been approved by `petrochenkov,ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov,ChrisDenton 2d25c3b3697728ea42eb26f79e6f80907470ca22 -->
<!-- homu: {"type":"Approved","sha":"2d25c3b3697728ea42eb26f79e6f80907470ca22","approver":"petrochenkov,ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

CKingX commented:
Thank you Chris, erikdesjardins and CryZe!
